internal/pkg/service/buffer/sink/tablesink/storage/volume/assigner/assigner.go

```diff
@@ -0,0 +1,177 @@
+package assigner
+
+import (
+	"math/rand"
+	"sort"
+	"strconv"
+
+	"github.com/cespare/xxhash/v2"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/sink/tablesink/storage"
+)
+
+// VolumesFor returns volumes list according to the file settings.
+// See assignVolumes function for details.
+func VolumesFor(volumes []storage.VolumeMetadata, file storage.File) ([]storage.VolumeMetadata, error) {
+	return assignVolumes(
+		volumes,
+		file.LocalStorage.Volumes.Count,
+		file.LocalStorage.Volumes.PreferredTypes,
+		file.OpenedAt().Time().UnixNano(),
+	), nil
+}
+
+// assignVolumes returns the requested number of volumes, if so many volumes are available.
+//
+// The "preferredTypes" slice defines priority of the volumes types.
+// The first value in the "preferredTypes" slice has the highest priority.
+// The function tries to assign the maximum number of volumes from the following levels:
+//   - the most preferred type, then
+//   - less preferred types, then
+//   - other types that are not on the preferred list
+//
+// Each selection is made from a different node, if possible.
+//
+// The "randomSeed" argument determines volume selection on the same priority level, and nodes order.
+func assignVolumes(input []storage.VolumeMetadata, limit int, preferredTypes []string, randomSeed int64) (out []storage.VolumeMetadata) {
```

Comment:
New implementation, main change is, that the main change is that we try to split the load to as many nodes as possible, while respecting preferredTypes.

One one can have mounted multiple volumes.